### DISCUSSION TITLE
problem about sw-django and sw-psycopg2

### DISCUSSION THREAD

**Discussion by Esdsnqxz** [apache/skywalking#8556] [Q&A] [answered]
when i use django as web framework with legency setup script of SW ,i get the following error
```
Traceback (most recent call last):
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\utils\autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\core\management\commands\runserver.py", line 121, in inner_run
    self.check_migrations()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\core\management\base.py", line 486, in check_migrations
    executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\migrations\executor.py", line 18, in __init__
    self.loader = MigrationLoader(self.connection)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\migrations\loader.py", line 53, in __init__
    self.build_graph()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\migrations\loader.py", line 220, in build_graph
    self.applied_migrations = recorder.applied_migrations()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\migrations\recorder.py", line 77, in applied_migrations
    if self.has_table():
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\migrations\recorder.py", line 55, in has_table
    with self.connection.cursor() as cursor:
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\utils\asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\backends\base\base.py", line 259, in cursor
    return self._cursor()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\backends\base\base.py", line 235, in _cursor
    self.ensure_connection()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\utils\asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\backends\base\base.py", line 219, in ensure_connection
    self.connect()
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\utils\asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\backends\base\base.py", line 200, in connect
    self.connection = self.get_new_connection(conn_params)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\utils\asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\django\db\backends\postgresql\base.py", line 206, in get_new_connection
    psycopg2.extras.register_default_jsonb(conn_or_curs=connection, loads=lambda x: x)
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\psycopg2\_json.py", line 152, in register_default_jsonb
    return register_json(conn_or_curs=conn_or_curs, globally=globally,
  File "C:\apps\anaconda\anzhuang\envs\dj\lib\site-packages\psycopg2\_json.py", line 120, in register_json
    register_type(JSON, not globally and conn_or_curs or None)
TypeError: argument 2 must be a connection, cursor or None
```

my SW legency setup script in urls.py as follow:
```
from skywalking import agent,config
config.init(collector_address="127.0.0.1:12800",service_name='ninjaserviceA',protocol='http')
config.django_collect_http_params=True
config.logging_level='DEBUG'
agent.start()
```
and the version of all 

- python : 3.8
- django ：3.2.5 
- psycopg2-binary : 2.9.3 
- SW-OAP: docker 8.9.0 

if i don't use skywlking the server can boot success

**Comment by kezhenxu94** [comment]
@Esdsnqxz did your program works normally if you remove the SkyWalking Agent? From the logs there is nothing related to SkyWalking Agent. If so, please provide a reproducible demo program

**Reply by Esdsnqxz** [reply]
> @Esdsnqxz did your program works normally if you remove the SkyWalking Agent? From the logs there is nothing related to SkyWalking Agent. If so, please provide a reproducible demo program

yes, if i remove SW agent, i can boot the server success, the django program as follow:

creat an app in django program,in the app 

models.py
```
from django.db import models

# Create your models here.
class Test(models.Model):
    id = models.SmallIntegerField(primary_key=True,verbose_name="id")
    x = models.CharField(max_length=100,null=False,verbose_name='x')
    y = models.CharField(max_length=100,null=False,verbose_name='y')

    class Meta:
        db_table = 'test'
        verbose_name = u'for test'

    def __str__(self):
        return self.id
```
urls.py
```
from django.urls import path,include
from app_1 import views

from skywalking import agent,config

config.init(collector_address="127.0.0.1:12800",service_name='ninjaService',protocol='http')
config.django_collect_http_params=True
config.logging_level='DEBUG'
agent.start()

urlpatterns = [
    path("test/get/",views.get,name='testapi-get'),
    path("test/post/", views.post, name='testapi-post'),
    path("test/delete/", views.delete, name='testapi-delete'),
]

```

views.py
```
from rest_framework import status
from rest_framework.response import Response
from app_1.models import Test
from app_1.serializer import TestSerializer
from rest_framework.decorators import api_view

def get_obj(pk_id):
    try:
        return Test.objects.get(id=pk_id)
    except Test.DoesNotExist:
        return


@api_view(["GET"])
def get(request):
    pk_id = request.GET.get('pk_id')
    obj = get_obj(pk_id)
    if not obj:
        return Response(data={'eml':'no record'},status=status.HTTP_404_NOT_FOUND)
    s = TestSerializer(instance=obj)
    return Response(data=s.data,status=status.HTTP_200_OK)
```
serializer.py
```
from rest_framework import serializers
from app_1.models import Test

class TestSerializer(serializers.ModelSerializer):
    class Meta:
        model = Test
        fields = '__all__'
```

apps.py

```
from django.apps import AppConfig

class App1Config(AppConfig):
    default_auto_field = 'django.db.models.BigAutoField'
    name = 'app_1'
```

admin.py

```
from django.contrib import admin

# Register your models here.
from app_1.models import Test

admin.site.register(Test)
```

in settings config pg database 

if i remove the SW agent ,i can boot the server ,and test api in postman 

```
System check identified no issues (0 silenced).
February 17, 2022 - 15:36:34
Django version 3.2.5, using settings 'settings.test'
Starting development server at http://127.0.0.1:8000/
Quit the server with CTRL-BREAK.
[17/Feb/2022 15:36:44] "GET /app_A/test/get/?pk_id=1 HTTP/1.1" 200 32
```

**Reply by Superskyyy** [reply]
@Esdsnqxz Hello, thank you for providing the demo, I will try to reproduce the problem later.

**Reply by Esdsnqxz** [reply]
thanks a lot for help , is there a problem in my demo or the configuration ? @kezhenxu94  @Superskyyy

**Reply by Superskyyy** [reply]
> thanks a lot for help , is there a problem in my demo or the configuration ? @kezhenxu94 @Superskyyy

Hello, would you mind pushing the demo code into a github repo, so I can test with it? @Esdsnqxz thanks.

**Reply by Superskyyy** [reply]
After some research over similar issues regarding Django and psycopg2, it is curious to see that Datadog and AWS x-ray also face this problem which is still likely unsolved, this will take a bit more time for me to investigate but hopefully dealt with soon. 
The root cause was brought by a change introduced in Django 3.1.1, we will need to find a good way to mitigate the issue.

Meanwhile, probably using Django 3.1 (any version <3.1.1) will serve as a temporary workaround, please try to downgrade Django to 3.1.0 if that is an option for you.  @Esdsnqxz 

ref links: `https://github.com/DataDog/dd-trace-py/issues/2438#issuecomment-906661786`
`https://github.com/aws/aws-xray-sdk-python/issues/243`

**Reply by Esdsnqxz** [reply]
> 

thanks again , i will try Djano 3.1.0.

**Reply by sphlvn3** [reply]
is there an update here?

**Reply by jonilsonds9** [reply]
Hi guys, I'm having the same problem but using Flask with postgres and psycopg2. Any updates?

**Reply by Superskyyy** [reply]
> Hi guys, I'm having the same problem but using Flask with postgres and psycopg2. Any updates?

I haven't seen a report on flask with psycopg2 or psycopg3 so far. Can you report the specific versions of flask and psycopg?

If you are in urgent need to make things work. Please either try downgrade flask(not sure the version as this was not reported previously) or disable the psycopg plugin in skywalking Python by supplying the agent config for now. 

I recently researched on the problem root solution but it's quite complicated, please bare with me until I implement the fix.

**Reply by jonilsonds9** [reply]
I'm using `psycopg2-binary==2.9.3` and `flask` in version `2.1.1`. I'm just starting to use python-agent. But I still haven't tried disabling plugin in skywalking Python.

But I'll wait for news here!

**Reply by Superskyyy** [reply]
> I'm using `psycopg2-binary==2.9.3` and `flask` in version `2.1.1`. I'm just starting to use python-agent. But I still haven't tried disabling plugin in skywalking Python.
> 
> 
> 
> But I'll wait for news here!

Thanks, please try either doable plugin or downgrade for the time being, whichever suits your needs. Then when I fix it you can start using the latest master branch. 

Please anyone above or following this thread I will ping you guys when fix is applied.

**Reply by bdronneau** [reply]
Hey,
Any news on this behavior ?
Best

### EDIT

Environment
- psycopg2==2.8.6
- djangorestframework==3.13.1
- python==3.6.13
- gunicorn

When disable `sw_psycopg2` the application working great but in trace of course DB statement missing

**Reply by Esdsnqxz** [reply]
> Hey, Any news on this behavior ? Best
> 
> ### EDIT
> Environment
> 
> * psycopg2==2.8.6
> * djangorestframework==3.13.1
> * python==3.6.13
> * gunicorn
> 
> When disable `sw_psycopg2` the application working great but in trace of course DB statement missing

hi，the cause of my problem is i use the wrong version of django.

**Comment by tom-pytel** [comment, marked as answer]
See https://github.com/apache/skywalking-python/pull/211

**Comment by Superskyyy** [comment]
@Esdsnqxz @sefaphlvn @jonilsonds9 @bdronneau FYI as you guys were troubled by this bug. ^

**Comment by bdronneau** [comment]
Thank, I will try to test it.

**Reply by Superskyyy** [reply]
> Thank, I will try to test it. Le ven. 3 juin 2022 à 20:28, Superskyyy ***@***.***> a écrit :
> […](#)
> @Esdsnqxz <[@Esdsnqxz](https://github.com/Esdsnqxz)> @sefaphlvn <[@sefaphlvn](https://github.com/sefaphlvn)> @jonilsonds9 <[@jonilsonds9](https://github.com/jonilsonds9)> @bdronneau <[@bdronneau](https://github.com/bdronneau)> FYI as you guys were troubled by this bug. ^ — Reply to this email directly, view it on GitHub <[#8556 (comment)](https://github.com/apache/skywalking/discussions/8556#discussioncomment-2880161)>, or unsubscribe <[github.com/notifications/unsubscribe-auth/AA72DLXZ2BXDHORBUK3KICTVNJFEHANCNFSM5OTZQTZQ](https://github.com/notifications/unsubscribe-auth/AA72DLXZ2BXDHORBUK3KICTVNJFEHANCNFSM5OTZQTZQ)> . You are receiving this because you were mentioned.Message ID: ***@***.***>

There should be an official release this week, or you can directly use master, it's tested. Welcome to provide feedback.

**Comment by ishallbethat** [comment]
I encounter the same error in flask "TypeError: argument 2 must be a connection, cursor or None".
If I set SW_AGENT_DISABLE_PLUGINS to sw_psycopg2 to disable the psycopg2 the error is gone and i don't see any database related trace reported to skywalking oap.

**Comment by Superskyyy** [comment]
> I encounter the same error in flask "TypeError: argument 2 must be a connection, cursor or None".
> 
> If I set SW_AGENT_DISABLE_PLUGINS to sw_psycopg2 to disable the psycopg2 the error is gone and i don't see any database related trace reported to skywalking oap.

It's fixed in the latest master, did you give it a try?

**Reply by ishallbethat** [reply]
I do not see it's released in https://pypi.org/project/apache-skywalking/. Latest is 0.7.0. 
I manually packaged and uploaded skywalking-python and used it in project. A error i got as below.
```
  File "/usr/local/lib/python3.7/site-packages/skywalking/agent/__init__.py", line 28, in <module>
    from skywalking.command import command_service
  File "/usr/local/lib/python3.7/site-packages/skywalking/command/__init__.py", line 18, in <module>
    from skywalking.command.command_service import CommandService
  File "/usr/local/lib/python3.7/site-packages/skywalking/command/command_service.py", line 21, in <module>
    from skywalking.protocol.common.Common_pb2 import Commands, Command
ModuleNotFoundError: No module named 'skywalking.protocol.common'
```
Can you enlighten me how to use the latest package properly ?

**Reply by Superskyyy** [reply]
> Reference in new issue
> 
> ### Reference in new issue
> Repository
> skywalking 
> Title
> 
> Body

Hi, a new release will come out within the week I suppose, before that to use it you will need to clone the repository submodule so we have the protocol to communicate with backend OAP. 

After you initialized the submodule, simply run `make install` , or you can package it to a wheel by `make package`
